File: java-streams/examples/java/src/main/java/streams/ChainList.java
package streams;

import java.util.List;
import java.util.stream.Collectors;

public class ChainList {

    List<Employee> findEmployees(List<Person> persons) {

        return persons.stream()
                .filter(person -> person.isEmployed())
                .map(person -> new Employee(person.getFirstName() + " " + person.getLastName()))
                .collect(Collectors.toList());
    }

    List<String> employeeNames(List<Employee> employees) {
        return employees.stream()
                .map(Employee::getName)
                .collect(Collectors.toList());
    }

    List<String> nameList(List<Person> persons) {
        return employeeNames(findEmployees(persons));
    }
}

File: java-streams/examples/kotlin/src/main/kotlin/streams/ChainList.kt
package streams

class ChainList {

    internal fun findEmployees(persons: List<Person>) =
            persons
                    .filter(Person::isEmployed)
                    .map { Employee(it.firstName + " " + it.lastName) }

    internal fun employeeNames(employees: List<Employee>) = employees.map { it.name }

    internal fun nameList(persons: List<Person>) = employeeNames(findEmployees(persons))
}
